1-Draft/RFCNNNN-Test-Connection.md

```diff
@@ -0,0 +1,125 @@
+---
+RFC: XXXX
+Author: Joel Sallow
+Status: Draft
+Version: 1.0
+Area: Cmdlets
+Comments Due: 6/10/19
+Plan to Implement: Yes
+---
+
+# Proposed Changes to `Test-Connection` Cmdlet
+
+## Syntax
+
+```
+# All Sets
+Test-Connection [-TargetName] <string[]> [-IPv4] [-IPv6] [-ResolveDestination] [-TimeoutSeconds <int>] [-Quiet] [<CommonParameters>]
+
+# Set 1
+[-Ping] [-Count <int>] [-Delay <int>] [-Source <string>] [-MaxHops <int>] [-BufferSize <int>] [-DontFragment]
+
+# Set 2
+[-Ping] [-Continues] [-Delay <int>] [-Source <string>] [-MaxHops <int>]  [-BufferSize <int>] [-DontFragment] 
+
+# Set 3
+Test-Connection [-TargetName] <string[]> -Traceroute [-Source <string>] [-MaxHops <int>] 
+
+# Set 4
+Test-Connection [-TargetName] <string[]> -MTUSizeDetect
+
+# Set 5
+Test-Connection [-TargetName] <string[]> -TCPPort <int> [-Source <string>]
+```
+
+## Changes
+
+1. All current host/information output is relegated to the `-Verbose` stream.
+2. No progress bar per committee decision in [PowerShell#6768](https://github.com/PowerShell/PowerShell/issues/6768).
+3. All output occurs as soon as data is available, individual records per ping reply / trace hop.
+4. All presented data should be readily available as properties to the user.
+
+## Output Type Proposed Members
+
+### PingStatus
+
+```csharp
+class PingStatus
+{
+    string Source { get; }
+    IPAddress Address { get; }
+    string Destination { get; }
+    int RoundtripTime { get; }
+    int BufferSize { get; }
+    PingOptions Options { get; }
+}
+
+// For -MtuSizeDetect switch
+class PingMtuDetect : PingStatus
+{
+    int MtuSize { get; }
+}
+```
+
+### TraceStatus
+
+```csharp
+class TraceStatus
+{
+    int Hop { get; }
+    int[] RoundTripTimes { get; }
+    string Destination { get; }
+    IPAddress DestinationAddress { get; }
+    IPAddress HopAddress { get; }
+    int BufferSize { get; }
+    PingStatus[] Replies { get; }
+}
+```
+
+## Mockups
+
+### `Test-Connection www.google.com`
+
+- Information from the `Replies` property of the output object should be included in the main output object, and primary output mode should be _multiple_ objects, each representing a single ping attempt/reply object.
+- Buffer _data_ is generally irrelevant, as it cannot be specified, and only a `BufferSize` property should be exposed. `Replies.Buffer` property should remain `private`.
+- `Replies.Options` property should be hidden from default formatting.
+- Results output as a table, grouped by Destination address (in the case that multiple destinations are specified).
+
+**Resulting output:**
+```code
+   Destination: www.google.com
+Source  Address       RoundtripTime BufferSize
```

Comment:
Should have units

Reply:
In the column names or the individual row formats? 

If we do it with column names, we have a potential for confusion where the column name no longer actually matches the property name. That's _probably_ a minor concern? I guess? Not sure.

If we do it with the formatting for the values in the columns themselves (e.g., `10 ms`) then they look like they might be strings rather than their actual data type.

Which is the more effective and least confusing approach here?

Reply:
I think it should be in the column name.  Agree that it can cause confusion, but if it's enclosed in parens, I think it's fairly obvious that's not part of the property name.  In the future, it would be nice if the formatter use coloring to indicate decoration that isn't part of the object.

Reply:
I think the one thing that bothers me about that is the wasted space. You end up with a column **18** characters wide (not including space before and after column) for values that are basically never more than 3 characters wide, at most, except in truly exceptional circumstances.

I'll update the RFC per your recommendation so you can see what I'm talking about.